doc: add description for filehandle.write(string[, position[, encoding]])

darahayes opened the pull request:
Hey folks, this PR adds missing docs for `filehandle.write(string[, position[, encoding]])` in the `fs.promises` API.

As per https://github.com/nodejs/node/issues/20406, only the buffer version of this API is documented right now.

One thing I wanted to ask. I believe the intention is to reconcile the Promise based API as much as possible with the callback based one. In the docs submitted with this PR, I described the `encoding` option has having the default value `utf8`. This is the behaviour in the callback based API. 

Based off the code [here](https://github.com/nodejs/node/blob/master/lib/internal/fs/promises.js#L252-L256). I don't think that's actually happening in the Promise based version.

I have two questions:

* Should the behaviour be modified such that a default value of `utf8` is used? I'd be delighted to provide a PR and some tests.
* Or should I just remove any mention of it from this PR?

Cheers!

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

vsemozhetbyt commented:
CI-lite: https://ci.nodejs.org/job/node-test-pull-request-lite-pipeline/1100/

vsemozhetbyt commented:
cc @jasnell

thefourtheye commented on doc/api/fs.md:
```diff
@@ -3918,6 +3918,33 @@ On Linux, positional writes do not work when the file is opened in append mode.
+the value will be coerced to one.
+
+`position` refers to the offset from the beginning of the file where this data
+should be written. If `typeof position !== 'number'` the data will be written at
```
Could this be simply written as, "If the type of position is not a number"?

darahayes replied:
So I basically copied the same wording used here: https://github.com/nodejs/node/pull/23224/files#diff-acabf706a8aa070a8796e3573f7e4678R3910

This style is used in the callback APIs too. I figured I would use the same wording that's used in those other places. I'd be happy to clarify those other places also if you think that's a good idea.

thefourtheye commented:
> Based off the code here. I don't think that's actually happening in the Promise based version.

Looks like you are correct. If possible, could you come up with a test case which will break because of this?

---

cc @nodejs/fs

darahayes commented:
>  If possible, could you come up with a test case which will break because of this?

Sure thing. I'll look to see how it's tested in the callback based API and see if I can get some inspiration from there.

benjamingr commented:
> Should the behaviour be modified such that a default value of utf8 is used? I'd be delighted to provide a PR and some tests.

I believe so - yes, and work improving the promises fs API is always appreciated. (cc @jasnell )

Changes in this PR LGTM

benjamingr commented on doc/api/fs.md:
```diff
@@ -3918,6 +3918,33 @@ On Linux, positional writes do not work when the file is opened in append mode.
+`encoding` is the expected string encoding.
+
+It is unsafe to use `filehandle.write()` multiple times on the same file
+without waiting for the `Promise` to be resolved (or rejected). For this
```
Nit: a rejected promise is still resolved (and so is a promise following another promise) - `fulfilled` instead of resolved or "settled" instead of either would work here.

darahayes replied:
Hey there, I've addressed some other feedback but I haven't addressed this one yet. I did that for two reasons.

1. This is the same wording used in other parts of the documentation. See the docs for the [filehandle.write(buffer)](https://github.com/nodejs/node/blob/1532232dd95f58378056864b6d1294d45b80c82d/doc/api/fs.md#filehandlewritebuffer-offset-length-position)
2. I would argue that explicitly using the words `resolved` and `rejected` make sense in this context. If you're familiar with Promises, this makes sense. I would argue the same cannot exactly be said about the word `fulfilled` because it's not as commonly used when referring to Promises. (at least as far as I can tell). In my opinion that potentially introduces some ambiguity.

All that being said: I am happy to make the suggested changes here, as well as in the other parts of the same doc to ensure consistency.

thefourtheye commented:
@darahayes I would say, you can add them as separate commits in this PR itself. Because without the code changes, this doc patch would not be correct.

ChALkeR commented on doc/api/fs.md:
```diff
@@ -3918,6 +3918,33 @@ On Linux, positional writes do not work when the file is opened in append mode.
+
+Write `string` to the file. If `string` is not a string, then
+the value will be coerced to one.
+
```
Misses the information about returned value in the resulting `Promise`.
See https://github.com/nodejs/node/blob/master/doc/api/fs.md#filehandlewritebuffer-offset-length-position for comparison:
> The `Promise` is resolved with an object containing…

ChALkeR replied:
Otherwise, looks good to me.

thefourtheye commented:
@darahayes No problem at all. I am sorry I couldn't get to this on time. Let me take a look at this today. Thanks for being patient.

thefourtheye commented:
Okay, looks like I was wrong.

> In the docs submitted with this PR, I described the encoding option has having the default value utf8. This is the behaviour in the callback based API.

> Based off the code here. I don't think that's actually happening in the Promise based version.

I took a deeper look at the code and the default encoding value is actually set to UTF-8. Let me break this down.

https://github.com/nodejs/node/blob/dde9b3a0d3a206bd43946b2087f0d39ccc967c48/lib/internal/fs/promises.js#L255-L256

which actually parses the encoding like this

https://github.com/nodejs/node/blob/6ae6383fdc40c1fe799a8ff600e60ed58366f798/src/node_file.cc#L1720

`args[3]` is actually the `length`. So, `ParseEncoding` does this

https://github.com/nodejs/node/blob/fcd7a7242890b30c3c143255234bdfe1855999b6/src/node_encoding.cc#L89-L100

If the `length` parameter is actually not a String then the default encoding (UTF8) will be used. Also, if the passed `length` value is not a valid encoding type, then

https://github.com/nodejs/node/blob/fcd7a7242890b30c3c143255234bdfe1855999b6/src/node_encoding.cc#L79-L85

will take care of it.

---

@darahayes Sorry, I misled you. Could you please drop the commit which introduces default value for encoding?

darahayes commented:
@thefourtheye thanks so much for your help! I'm kicking myself that I didn't figure that out sooner. I've dropped the commit you mentioned and I also addressed some of the feedback on the docs changes.

thefourtheye commented:
Lite-CI: https://ci.nodejs.org/job/node-test-pull-request-lite-pipeline/1612/

thefourtheye commented:
@nodejs/fs PTAL once.

Trott commented:
Landed in 692b09ffbe5155
